---
title: "DanubeSedimentChlamydiae"
author: "AHorn"
date: "3/8/2022"
output: html_document
---

Using ASV data from the Danube sediment experiment (T. Halter, A. Horn) generated by the JMF.

Analyses based on [Callahan et al, 2017](https://bioconductor.org/help/course-materials/2017/BioC2017/Day1/Workshops/Microbiome/MicrobiomeWorkflowII.html), [Lahti et al](https://microbiome.github.io/tutorials/Composition.html), [ANCOM-BC](https://github.com/FrederickHuangLin/ANCOM-BC-Code-Archive)
[r-from-a-learners-perspective](https://r-from-a-learners-perspective.readthedocs.io/en/latest/part5/)

in order to be able to get fna file into R, the fasta header need to be shortened:
  `sed '/^>/ s/ .*//' DADA2_ASVs.fna > DADA2_ASVs_short.fna`

## Calculate basic phylogenetic tree

Can be done in R but for larger datasets on a laptop it is not feasible to calculate the tree.
Thus, the alignment was performed with mafft by adding the short sequences to the existing whole 16S alignment for chlamydiae: `mafft --addfragments DADA2_ASVs_short.fna --reorder --keeplength --6merpair --thread -1 Chlamydiae_GEM_Chls_Silva138_ali_full_drep99_ali.fasta > Chlamydiae_GEM_Chls_Silva138_with_DanubeSedASVs.fasta`

ASVs were extracted from large alignment and only ASVs used to calculate the Fasttree.
`FastTree -nt -gtr DADA2_ASVs_aln2_Chlamydiae_GEM_Chls_Silva138.fasta > DanubeSedASVs_fasttree.nwk`

## Generate dataset for further analysis
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# load neccessary packages
library(phyloseq)
library(ape)
library(phytools)
library(plyr)
library(ggplot2)
library(gridExtra)
library(vegan)
library(magrittr)
library(ANCOMBC)
library(tidyr)
library(RColorBrewer)
library(microViz)

```

```{r generate dataset, include=T}
# 1. Import all necessary files
OTU <- read.table("Chl_DADA2_counts_as_matrix.tsv", header=TRUE, row.names=1, check.names=F)
OTU <- as.matrix(OTU, rownames = 1)

TAX <- read.table("DADA2_ASVs.rRNA_SSU.custom_reference.DADA2_classified.tsv", sep = "\t", header = T, row.names=1, na.strings=c("","NA"))
TAX <- as.matrix(TAX)

Sample <- read.table("metadata_Chl.csv", sep="\t", header=TRUE, row.names=1)

Tree <- read.tree("DanubeSedChl_Fasttree.nwk")
#Tree_mid <- midpoint.root(Tree) # tree needs to be rooted for later analysis
Tree_rooted <- root(Tree, "ASV_t9d_ru3", resolve.root=T) # root tree at one of the ASVs that is in the outgroup when adding all ASVs to the chlamydial tree with epa-ng
is.rooted(Tree_rooted)
Tree_bifur <- ape::multi2di(Tree_rooted) # tree needs to have bifurcations only for wunifrac

# 2. Generate phyloseq object
PS_all <- phyloseq(otu_table(OTU, taxa_are_rows=T), tax_table(TAX), sample_data(Sample), phy_tree(Tree_bifur))
PS_all

#correct taxonomy - replace NA with last valid taxonomic rank
PS_all <- tax_fix(
  PS_all,
  min_length = 4,
  unknowns = NA,
  suffix_rank = "classified",
  sep = " ",
  anon_unique = TRUE,
  verbose = TRUE
)

# 3. Apply different filtering steps
# Remove negative control, positive control, TP1
PS <- prune_samples(sample_names(PS_all) != c('JMF-1904-4-0001F','JMF-1904-4-0007F','JMF-1904-4-0042F'), PS_all) 
PS
# filter out samples with no abundance
PS <- prune_samples(sample_sums(PS)>0, PS) 

# Include only taxa with more than 20 reads (on average) in at least 5% of samples
filter <- phyloseq::genefilter_sample(PS, filterfun_sample(function(x) x >= 20))#, A = 0.05*nsamples(PS)) # if also 0.05% of samples included - would leave 218 taxa
PS2 <- prune_taxa(filter, PS)
PS2

# order samples by season
correct.order <- c("spring","summer","autumn","winter")
sample_data(PS2)$Season <- factor(sample_data(PS2)$Season, levels = correct.order)
sample_data(PS)$Season <- factor(sample_data(PS)$Season, levels = correct.order)
# order samples by season and year
correct.order2 <- c("spring1","summer1","autumn1","winter1","spring2","summer2","autumn2","winter2","spring3")
sample_data(PS2)$SeasonYear <- factor(sample_data(PS)$SeasonYear, levels = correct.order2)
sample_data(PS)$SeasonYear <- factor(sample_data(PS)$SeasonYear, levels = correct.order2)
# order months in correct way
correct.order3 <- c("March","April","May","June","July","August","September","October","November","December","January","February")
sample_data(PS2)$Month <- factor(sample_data(PS2)$Month, levels = correct.order3)
sample_data(PS)$Month <- factor(sample_data(PS)$Month, levels = correct.order3)


# get basic overview over data and composition

reads_samples <- data.frame(sample_sums(PS)) #read sums per sample
reads_total <- sum(reads_samples)
reads_ASV <- data.frame(taxa_sums(PS)) #reads

reads_samplesfilt <- data.frame(sample_sums(PS2)) #read sums per sample
reads_total_filt <- sum(reads_samplesfilt)
reads_ASVfilt <- data.frame(taxa_sums(PS2)) #reads

#check reads in samples before and after filtering
plot_bar(PS, fill="Family")
#plot_bar(PS2, fill="Family")

#removed sample 41 for eukaryotes, because of low read nr - also remove it here!

# alpha diversity - use unfiltered data because of singletons
col_season = c("spring"="darkseagreen3","summer" = "deeppink3","autumn" ="goldenrod1","winter" = "deepskyblue3")

plot_richness(PS, measures=c("Observed","Chao1","Shannon")) + geom_boxplot()
plot_richness(PS, x="Month", measures=c("Observed", "Chao1","Shannon","InvSimpson")) + 
  geom_boxplot() 
ggsave("alpha_div_month_chl.png")
# Feb and June have two outliers for chlamydiae...
plot_richness(PS, color = "Season", x="Season", measures=c("Observed","Chao1","InvSimpson")) + #"Shannon",
  geom_boxplot() +
  scale_color_manual(values = col_season)
ggsave("alpha_div_season_chl.png")

# rarefaction curve
rarecurve(t(otu_table(PS2)), step=50, cex=0.5)
ggsave("rarefaction_curve_chl.png")
# rarefaction curve looks way better than for euks

# remove sample 41 from further analysis - as in euk analysis
PS2 <- prune_samples(sample_names(PS2) != c('JMF-1904-4-0041F'), PS2) 

reads_samplesPS2 <- data.frame(sample_sums(PS2)) #read sums per sample
reads_total_PS2 <- sum(reads_samplesPS2)
reads_ASVPS2 <- data.frame(taxa_sums(PS2)) #reads
# total chlamydial reads
#originally
reads_total <- sum(taxa_sums(PS_all))
#used reads
reads_used <- sum(taxa_sums(PS2))

########
#change to taxonomy derived by epa-ng placement of ASVs

```

## Switch to analysis based on taxonomy based on tree (not blast)

```{r TAX tree analyses, include=T}
######TAX tree analyses!!!!!
#export PS2 ASVs
PS2_ASVs <- otu_table(PS2)
write.csv(PS2_ASVs, "PS2_ASVs.csv")

# generate novel PS object with taxonomy derived from epa-ng tree (slightly different and more ASVs with affiliation)

TAXtree <- read.table("chl_PS2_treetax.tsv", sep = "\t", header = T, row.names=1, na.strings=c("","NA"))
# replace na by none, to have those in later analyses too
TAXtree$Family <- TAXtree$Family %>% replace_na("no_family")
TAXtree$Genus <- TAXtree$Genus %>% replace_na("no_genus")
TAXtree <- as.matrix(TAXtree)
Sample_2 <- sample_data(PS2)

PS2_taxtree <- phyloseq(otu_table(PS2_ASVs, taxa_are_rows=T), tax_table(TAXtree), sample_data(Sample_2), phy_tree(phy_tree(PS2)))
PS2_taxtree

#check reads in samples
reads_samples_used <- data.frame(sample_sums(PS2_taxtree)) #read sums per sample
reads_total_used <- sum(reads_samples_used)
reads_ASVs_used <- data.frame(taxa_sums(PS2_taxtree)) #reads
plot_bar(PS2_taxtree, fill="Family")

# order samples by season
correct.order <- c("spring","summer","autumn","winter")
sample_data(PS2_taxtree)$Season <- factor(sample_data(PS2_taxtree)$Season, levels = correct.order)
# order samples by season and year
correct.order2 <- c("spring1","summer1","autumn1","winter1","spring2","summer2","autumn2","winter2","spring3")
sample_data(PS2_taxtree)$SeasonYear <- factor(sample_data(PS2_taxtree)$SeasonYear, levels = correct.order2)
# order months in correct way
correct.order3 <- c("March","April","May","June","July","August","September","October","November","December","January","February")
sample_data(PS2_taxtree)$Month <- factor(sample_data(PS2_taxtree)$Month, levels = correct.order3)

#RA PS2_taxtree
PS2_taxtree_ra = PS2_taxtree %>% tax_glom(taxrank = "Family") %>% transform_sample_counts(function(x) {x/sum(x)}) %>% psmelt() %>% arrange(Family)
# order of Family levels
correct.order4 <- c("Simkaniaceae", "Rhabdochlamydiaceae","SM23-39", "Parachlamydiaceae", "Waddliaceae", "Criblamydiaceae", "PCF5", "group_7", "group_11", "GCA-2709385", "PCF7", "MCF-F", "group_2", "K940_chlam_8", "MCF-A", "MCF-E", "PCF1", "no_family")
PS2_taxtree_ra$Family <- factor(PS2_taxtree_ra$Family, levels = correct.order4)

#sort timepoints
correct.order5 <- c(39:2)

PS2_taxtree_ra$Timepoint <- factor(PS2_taxtree_ra$Timepoint, levels = correct.order5)

#define colors for family levels
my_color2 = c(Simkaniaceae="#bc80bd",Rhabdochlamydiaceae = "#bebada","SM23-39" = "#80b1d3",Parachlamydiaceae = "#8dd3c7",Waddliaceae="#aafae7",Criblamydiaceae="#ccebc5",PCF5="#b3de69",group_7="#ffffb3",group_11="#ffed6f","GCA-2709385" = "#fdb462",PCF7 = "#fb8072","MCF-F"="#fccde6",group_2="#d9d9d9","K940_chlam_8"="#969696","MCF-A"="#969696","MCF-E"="#969696","PCF1"="#969696",no_family = "#f2f2f2")

my_color2b = c(Simkaniaceae="#685582",Rhabdochlamydiaceae = "#90709a","SM23-39" = "#ceb7cd",Parachlamydiaceae = "#f9d9ed",Waddliaceae="#b99158",Criblamydiaceae="#f5daa3",PCF5="#fff397",group_7="#dbdfa2",group_11="#b7d968","GCA-2709385" = "#68a98f",PCF7 = "#8fbcc5","MCF-F"="#6d89a5",group_2="#4b5172","K940_chlam_8"="#969696","MCF-A"="#969696","MCF-E"="#969696","PCF1"="#969696",no_family = "#f2f2f2")

P_ra <- ggplot(PS2_taxtree_ra, aes(x = Timepoint, y = Abundance, fill = Family)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values= my_color2) +
  theme(axis.text.x = element_text(angle = 45, hjust =1)) +
  theme_light() +
  facet_wrap(SeasonYear~.,
                scales="free_y",
                strip.position="right",
                nrow=8
                ) +
  coord_flip() +
  xlab("") +
  ylab("rel abundance") 
ggsave("chl_rel_ab.png", width = 10, height = 15)

#### absolute abundance with ddPCR data
#add column in df containing rel abundance * chlamydial copies/g soil in that sample
PS2_taxtree_ab <- PS2_taxtree_ra %>% mutate(absnr = Abundance * cp_g_sediment)

#write.csv(PS2_taxtree_ab, "abs_ab_PS2_table.csv")

PS2_taxtree_ab$Timepoint <- factor(PS2_taxtree_ab$Timepoint, levels = correct.order5)
library(ggbreak)
P_absnr <- ggplot(PS2_taxtree_ab, aes(x = Timepoint, y = absnr, fill = Family)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values= my_color2) +
  #theme(axis.text.x = element_text(angle = 45, hjust =1)) +
  theme_light() +
  theme(axis.text.y = element_text()) +
  xlab("") +
  ylab("16S rDNA copies/g sediment") +
  facet_wrap(SeasonYear~.,
                scales="free_y",
                strip.position="right",
                nrow=8
                ) +
  coord_flip()
  
#P_absnr2 <- P_absnr +
#scale_y_cut(breaks=c(2600000), which=c(1,2), scales=c(3, 0.5))

P_absnr3 <- P_absnr +
scale_y_break(c(2700000, 3000000), scale=0.2)
#print(P_absnr3)
ggsave("Chl_absnr_2.png", width =10, height =10)

## same plot in high
P_absnr_high <- ggplot(PS2_taxtree_ab, aes(x = Timepoint, y = absnr, fill = Family)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values= my_color2) +
  theme_light() +
  xlab("") +
  ylab("16S rRNA gene copies/g sediment") +
  theme(axis.text.x = element_blank(),
       axis.text.y = element_text(size = 12),
       axis.title = element_text(size =16)) +
  facet_wrap(SeasonYear~.,
                scales="free_x",
                strip.position="top",
                ncol=8
                ) +
  scale_x_discrete(limits = rev)
  
P_absnr3_high <- P_absnr_high +
scale_y_break(c(2700000, 3000000), scale=0.2)
#print(P_absnr3)
ggsave("Chl_absnr_2_high.png", width =12, height =10)


#30 most abundant taxa
most_abundant_taxa30 <- sort(taxa_sums(PS2_taxtree), T)[1:30]
print(most_abundant_taxa30)
write.csv(most_abundant_taxa30, "most_abundant_taxa30.csv")

PS_top30 <- prune_taxa(names(most_abundant_taxa30), PS2_taxtree)

PS_top30_ra = PS_top30 %>% tax_glom(taxrank = "Family") %>% transform_sample_counts(function(x) {x/sum(x)}) %>% psmelt() %>% arrange(Family)
#reorder family and timepoints
PS_top30_ra$Family <- factor(PS_top30_ra$Family, levels = correct.order4)
PS_top30_ra$Timepoint <- factor(PS_top30_ra$Timepoint, levels = correct.order5)

P30ra <- ggplot(PS_top30_ra, aes(x = Timepoint, y = Abundance, fill = Family)) +
  geom_bar(stat = "identity") +
  theme(axis.title.x = element_blank()) +
  scale_fill_manual(values= my_color2) +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  theme_minimal() +
  xlab("") +
  ylab("rel abundance") 

#plot pH, watertemp, DOC, watercond
PS_top30_ra$Timepoint <- as.numeric(as.character(PS_top30_ra$Timepoint))
pH <- ggplot(PS_top30_ra, aes(x = Timepoint, y = pH)) +
  geom_line(color = "orange") +
  geom_point(aes(group=1), color = "orange") +
  theme(axis.title.x = element_blank()) +
  theme_minimal() +
  xlab("")
T <- ggplot(PS_top30_ra, aes(x = Timepoint, y = Watertemp_C)) +
  geom_line(color = "blue") +
  geom_point(aes(group=1), color = "blue") +
  theme(axis.title.x = element_blank()) +
  theme_minimal() +
  xlab("") +
  ylab("Water temperature [°C]")

DO <- ggplot(PS_top30_ra, aes(x = Timepoint, y = DO_mgperl)) +
  geom_line(color = "lightblue") +
  geom_point(aes(group=1), color = "lightblue") +
  theme(axis.title.x = element_blank()) +
  theme_minimal() +
  xlab("") +
  ylab("DO [mg/l]")

Con <- ggplot(PS_top30_ra, aes(x = Timepoint, y = Conductivity_µS)) +
  geom_line(color = "purple") +
  geom_point(aes(group=1), color = "purple") +
  theme(axis.title.x = element_blank()) +
  theme_minimal() +
  xlab("") +
  ylab("Conductivity [µS]")

pH_T <- ggplot(PS_top30_ra, aes(x = Timepoint)) +
  geom_line(aes(y = Watertemp_C, color = "Water T")) +
  geom_line(aes(y = pH*2, color = "pH")) +
  scale_y_continuous(sec.axis = sec_axis(~./2, name = "pH")) +
  scale_color_manual(values = c("orange", "blue")) +
  labs(y="Water temperature [°C]", x= "", colour ="") +
  theme_light() +
  scale_x_continuous(limits = c(2,39)) +
  expand_limits(x =2, y = 0) +
  theme(legend.position = c(0.9,0.9))

library(ggpubr)
ggarrange(P_ra, P30ra, ncol=1, nrow=2, common.legend = TRUE, legend="right")

ggarrange(pH, T, DO, Con, ncol=1, nrow=4)
ggsave("environmental_parameter.pdf", width = 15, height = 10)

ggarrange(pH, T, P30ra, ncol=1, nrow=3, align = c("v"), heights = c(0.25, 0.25, 0.5), legend = F)
ggsave("pH_T_top30chl.pdf", width = 15, height = 10)

ggarrange(pH_T, P30ra, ncol=1, nrow=2, align = "v", heights = c(0.25, 0.75), legend = FALSE)
ggsave("pHT_top30chl.pdf", width = 15, height = 10)

ggarrange(pH_T, P_ra, ncol=1, nrow=2, align = "v", heights = c(0.25, 0.75), legend = FALSE)
ggsave("pHT_chl_ra.pdf", width = 15, height = 10)

ggarrange(P_ra, P_absnr, ncol=1, nrow=2, align = "v", common.legend=TRUE, legend = "left")
ggsave("chl_rel_and_abs.pdf", width =15, height = 15)

ggarrange(print(P_ra), print(P_absnr3), ncol = 2, nrow =1, align = "h", common.legend=TRUE)#, align = "hv", legend = "bottom")
ggsave("chl_rel_and_abs_h.png", width = 20, height = 10)

```

## Differential abundance of taxa between seasons

```{r, ancomBC with taxtree data, include=T}
# ANCOM-BC with taxtree data

library(microbiome)
library(nloptr)
library(tibble)
library(DT)

### spring- summer
PS_springsummer <- subset_samples(PS2_taxtree, Season=="spring"|Season=="summer")

out_SS <- ancombc(phyloseq = PS_springsummer, formula = "Season",  group = "Season", neg_lb = T, conserve = TRUE, global = T)#struc_zero = T,

SS_res <- data.frame(row.names = row.names(out_SS$res$beta),
  beta = unlist(out_SS$res$beta),
  se = unlist(out_SS$res$se),
  W = unlist(out_SS$res$W),
  p_val = unlist(out_SS$res$p_val),
  q_val = unlist(out_SS$res$q_val),
  diff_abn = unlist(out_SS$res$diff_abn))
SS_ancom_fdr <- SS_res %>% dplyr::filter(q_val < 0.05)
SS_ancom_logfold <- SS_ancom_fdr %>% dplyr::filter(beta < -2.0 | beta > 2.0)
### summer -autumn
PS_summerautumn <- subset_samples(PS2_taxtree, Season=="summer"|Season=="autumn")
out_SA <- ancombc(phyloseq = PS_summerautumn, formula = "Season", group = "Season", neg_lb = T, conserve = TRUE, global = T)

SA_res <- data.frame(row.names = row.names(out_SA$res$beta),
  beta = unlist(out_SA$res$beta),
  se = unlist(out_SA$res$se),
  W = unlist(out_SA$res$W),
  p_val = unlist(out_SA$res$p_val),
  q_val = unlist(out_SA$res$q_val),
  diff_abn = unlist(out_SA$res$diff_abn))
SA_ancom_fdr <- SA_res %>% dplyr::filter(q_val < 0.05)
SA_ancom_logfold <- SA_ancom_fdr %>% dplyr::filter(beta < -2.0 | beta > 2.0)
### autumn - winter
PS_autumnwinter <- subset_samples(PS2_taxtree, Season=="autumn"|Season=="winter")
out_AW <- ancombc(phyloseq = PS_autumnwinter, formula = "Season", group = "Season", neg_lb = T, conserve = TRUE, global = T)

AW_res <- data.frame(row.names = row.names(out_AW$res$beta),
  beta = unlist(out_AW$res$beta),
  se = unlist(out_AW$res$se),
  W = unlist(out_AW$res$W),
  p_val = unlist(out_AW$res$p_val),
  q_val = unlist(out_AW$res$q_val),
  diff_abn = unlist(out_AW$res$diff_abn))
AW_ancom_fdr <- AW_res %>% dplyr::filter(q_val < 0.05)
AW_ancom_logfold <- AW_ancom_fdr %>% dplyr::filter(beta < -2.0 | beta > 2.0)
### winter - spring
PS_winterspring <- subset_samples(PS2_taxtree, Season=="winter"|Season=="spring")
out_WS <- ancombc(phyloseq = PS_winterspring, formula = "Season", group = "Season", neg_lb = T, conserve = TRUE, global = T)

WS_res <- data.frame(row.names = row.names(out_WS$res$beta),
  beta = unlist(out_WS$res$beta),
  se = unlist(out_WS$res$se),
  W = unlist(out_WS$res$W),
  p_val = unlist(out_WS$res$p_val),
  q_val = unlist(out_WS$res$q_val),
  diff_abn = unlist(out_WS$res$diff_abn))
WS_ancom_fdr <- WS_res %>% dplyr::filter(q_val < 0.05)
WS_ancom_logfold <- WS_ancom_fdr %>% dplyr::filter(beta < -2.0 | beta > 2.0)

# make a df with all taxa and beta (log fold change) data
SS_SA_merge <- merge(SS_res, SA_res, by = 0, all.x=TRUE, all.y=TRUE)
merge1 <- data.frame(SS_SA_merge$beta.x, SS_SA_merge$beta.y, row.names=SS_SA_merge$Row.names)
merge1_AW <- merge(merge1, AW_res, by = 0, all =TRUE)
merge2 <- data.frame(merge1_AW$SS_SA_merge.beta.x,merge1_AW$SS_SA_merge.beta.y,merge1_AW$beta, row.names=merge1_AW$Row.names)
merge2_WS <- merge(merge2, WS_res, by = 0, all =TRUE)
merge_all <- data.frame(merge2_WS$merge1_AW.SS_SA_merge.beta.x,merge2_WS$merge1_AW.SS_SA_merge.beta.y,merge2_WS$merge1_AW.beta,merge2_WS$beta, row.names=merge2_WS$Row.names)

# replace na with 0
merge_all[is.na(merge_all)] <- 0
# change col names
colnames(merge_all) <- c("SS","SA","AW","WS")
# filter only log fold change above 2
library(dplyr)
ancom_log_change <- merge_all %>% dplyr::filter_all(any_vars(. < -2.0 | . > 2.0))
# merge with tax data
ancom_log_change_tax <- merge(ancom_log_change, tax_table(PS2_taxtree), by=0, all.x=TRUE)

#plot log change as heatmap of log-fold over 2 changing ASVs
library(pheatmap)

###change order of comparisons - WS first
merge_all_WSfirst <- merge_all %>% relocate("WS")

colnames(merge_all_WSfirst) <- c("WS","SS","SA","AW")
# filter only log fold change above 2
ancom_log_change_WSfirst <- merge_all_WSfirst %>% dplyr::filter_all(any_vars(. < -2.0 | . > 2.0))
# merge with tax data
ancom_log_change_tax_WSfirst <- merge(ancom_log_change_WSfirst, tax_table(PS2_taxtree), by=0, all.x=TRUE)

#make matrix
matrix_change_WSfirst <- as.matrix(ancom_log_change_WSfirst)
#define row annotations
annotation_row_WSfirst = data.frame(
    Family = as.factor(ancom_log_change_tax_WSfirst$Family), row.names = ancom_log_change_tax_WSfirst$Row.names)
# define column annotations
my_columns_WSfirst <- data.frame(Season = rep(c("winter-spring","spring-summer","summer-autumn","autumn-winter")))
row.names(my_columns_WSfirst) <- colnames(ancom_log_change_WSfirst)
#define annotation colors
my_color_WSfirst = list(
  Family = c(Simkaniaceae="#bc80bd",Rhabdochlamydiaceae = "#bebada","SM23-39" = "#80b1d3",Parachlamydiaceae = "#8dd3c7",Waddliaceae="#aafae7",Criblamydiaceae="#ccebc5",group_7="#ffffb3",group_11="#ffed6f","GCA-2709385" = "#fdb462",PCF7 = "#fb8072",group_2="#d9d9d9","MCF-E"="#969696",no_family = "#f2f2f2"),
  Season = c("winter-spring"="darkseagreen3","spring-summer" = "deeppink3","summer-autumn" ="goldenrod1","autumn-winter" = "deepskyblue3")
)
#red family in legend for ISME plot
my_color_WSfirst_redfam = list(
  Family = c(Rhabdochlamydiaceae = "#bebada","SM23-39" = "#80b1d3",Parachlamydiaceae = "#8dd3c7",Waddliaceae="#aafae7",PCF7 = "#fb8072",group_2="#d9d9d9",no_family = "#f2f2f2"),
  Season = c("winter-spring"="darkseagreen3","spring-summer" = "deeppink3","summer-autumn" ="goldenrod1","autumn-winter" = "deepskyblue3")
)

heatmap_chl_WSfirst <- pheatmap(matrix_change_WSfirst, annotation_row = annotation_row_WSfirst, annotation_col = my_columns_WSfirst, annotation_colors=my_color_WSfirst_redfam,cluster_cols=F)
heatmap_chl_WSfirst

save_pheatmap_png <- function(x, filename, width=1200, height=1000, res = 150) {
  png(filename, width = width, height = height, res = res)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}

save_pheatmap_png(heatmap_chl_WSfirst, "heatmap_chl_taxtree_WSfirst.png")


#plot log change as heatmap of log-fold above 1 changing ASVs
# filter only log fold change above 1
ancom_log_change1 <- merge_all %>% dplyr::filter_all(any_vars(. < -1.0 | . > 1.0))
# merge with tax data
ancom_log_change_tax1 <- merge(ancom_log_change1, tax_table(PS2_taxtree), by=0, all.x=T)

#plot log change as heatmap of log-fold above 1 changing ASVs with WS first
# filter only log fold change above 1
ancom_log_change1_WSfirst <- merge_all_WSfirst %>% dplyr::filter_all(any_vars(. < -1.0 | . > 1.0))
# merge with tax data
ancom_log_change_tax1_WSfirst <- merge(ancom_log_change1_WSfirst, tax_table(PS2_taxtree), by=0, all.x=TRUE)

#make matrix
matrix_change1_WSfirst <- as.matrix(ancom_log_change1_WSfirst)
#define row annotations
annotation_row1_WSfirst = data.frame(
    Family = as.factor(ancom_log_change_tax1_WSfirst$Family), row.names = ancom_log_change_tax1_WSfirst$Row.names)
unique(ancom_log_change_tax1_WSfirst$Family)
#define annotation colors
my_color1_WSfirst = list(
  Family = c(Simkaniaceae="#bc80bd",Rhabdochlamydiaceae = "#bebada","SM23-39" = "#80b1d3",Parachlamydiaceae = "#8dd3c7",Waddliaceae="#aafae7",Criblamydiaceae="#ccebc5",group_7="#ffffb3",group_11="#ffed6f","GCA-2709385" = "#fdb462",PCF7 = "#fb8072",group_2="#d9d9d9","MCF-E"="#969696",no_family = "#f2f2f2"),
  Season = c("winter-spring"="darkseagreen3","spring-summer" = "deeppink3","summer-autumn" ="goldenrod1","autumn-winter" = "deepskyblue3")
)

heatmap_chl_log1_WSfirst <- pheatmap(matrix_change1_WSfirst, annotation_row = annotation_row1_WSfirst, annotation_col = my_columns_WSfirst, annotation_colors=my_color1_WSfirst,cluster_cols=F)
heatmap_chl_log1_WSfirst

save_pheatmap_png <- function(x, filename, width=1200, height=2000, res = 150) {
  png(filename, width = width, height = height, res = res)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}
save_pheatmap_png(heatmap_chl_log1_WSfirst, "heatmap_chl_log1_taxtree_WSfirst.png")


```
## redo all analyses from beginning and change to taxonomy derived by epa-ng placement of ASVs

```{r, redo analyses with taxonomy from epa-ng placement, include = FALSE}

# 4. Do more filtering
# Create table, number of features for each Family
table(tax_table(PS2_taxtree)[, "Family"], exclude = NULL)
## family with most taxa is now Para and not no_family any more (as with taxonomy from classification file)

# Phylogenetic diversity - Faith's phylogenetic distance
library(metagMisc)
#library(PhyloMeasures)

#FaithPD <- phyloseq_phylo_div(PS2_taxtree, measures = "PD")
# combine data with sample data
#sample_data2 <- data.frame(sample_data(PS2_taxtree))
#sample_data3 <- data.frame(FaithPD, sample_data2)
 
# my_comparisons <- list( c("spring","summer"), c("summer","autumn"), c("autumn","winter"), c("winter","spring"))
# my_comparisons2 <- list( c("spring1","spring2"), c("summer1","summer2"), c("autumn1","autumn2"), c("winter1","winter2"))
# pd_season <- ggboxplot(sample_data3, "Season", "PD", color="Season", add = "jitter") +
#   stat_compare_means(comparison = my_comparisons, label = "p.signif", hide.ns=T, step.increase=1) +
#   stat_compare_means(label.y = 45)
# pd_seasonyear <- ggboxplot(sample_data3, "SeasonYear", "PD", color="Season", add = "jitter") +
#   stat_compare_means(comparison = my_comparisons2, label = "p.signif", hide.ns=T, step.increase=1) +
#   stat_compare_means(label.y = 45)
# pd_winter <- ggboxplot(sample_data3, "Season", "PD", color="Season", add = "jitter") +
#   stat_compare_means(comparison = my_comparisons3, label = "p.signif", hide.ns=T, step.increase=1) +
#   stat_compare_means(label.y = 45)
# grid.arrange(pd_season, pd_seasonyear, pd_winter, ncol=2, nrow=2)
# #no significance concerning PD between seasons and years

```

## Beta diversity and ordination
```{r Ordination, include=T}
# Ordination plots
# do first ordination plot with log transformed (for normalisation - not ideal way of normalising!) data to check again for outliers

# PCoA plot with Bray-Curtis distance
pslog <- transform_sample_counts(PS2_taxtree, function(x) log(1 + x)) #normalise by log transformation
psclr <- microbiome::transform(PS2_taxtree, "clr") #normalise by centered log transformation
Season_col <- c("darkseagreen3","deeppink3","goldenrod1","deepskyblue3")

out.pcoa.log <- ordinate(pslog,  method = "MDS", distance = "bray")
evals <- out.pcoa.log$values[,1]
plot_ordination(pslog, out.pcoa.log, color = "Season") +
  labs(col = "Season") +
  theme_classic() +
  stat_ellipse(aes(group = Season)) +
  scale_color_manual(values = Season_col) +
  coord_fixed(sqrt(evals[2] / evals[1]))	
ggsave("PCoA__Bray_chl.png")
#PerMANOVA to check for significance
adonis(evals ~ sample_data(pslog)$Season)

# do MDS plot with log 
# out.uf.log <- ordinate(pslog, method = "MDS", distance = "unifrac")
# evals <- out.uf.log$values$Eigenvalues
# plot_ordination(pslog, out.uf.log, color = "Season") +
#   labs(col = "Season") +
#   theme_classic() +
#   stat_ellipse(aes(group = Season)) +
#   coord_fixed(sqrt(evals[2] / evals[1]))

out.wuf.log <- ordinate(pslog, method = "MDS", distance = "wunifrac")
evals <- out.wuf.log$values$Eigenvalues
plot_ordination(pslog, out.wuf.log, color = "Season") +
  labs(col = "Season") +
  theme_classic() +
  stat_ellipse(aes(group = Season)) +
  coord_fixed(sqrt(evals[2] / evals[1]))

p3 <- plot_scree(out.wuf.log, "wunifrac")
#p2 <- plot_scree(out.uf.log, "unifrac")
p1 <- plot_scree(out.pcoa.log, "bray")
ggsave("scree_plot_bray.png")
#grid.arrange(p1, p2, p3, ncol=3)
grid.arrange(p1, p3, ncol=2)

#####on bray distance and log transformed data!
### PERMANOVA - hypothesis testing
# PS2_otu_log <- as.matrix(otu_table(pslog))
# PS2_otu_log_dist <- phyloseq::distance(PS2_otu_log, method = "bray")
# PS2_data <- data.frame(sample_data(PS2))
# 
# adonis_season <- adonis(PS2_otu_log_dist ~ Season + SeasonYear + pH + Watertemp_C, data = PS2_data)
# adonis_season #season is very significant and seasonyear significant
# 
# ## anosim - analysis of similarities - test if env factors have influence on community composition
# ano_season <- anosim(PS2_otu_log_dist, PS2_data$Season)
# ano_season #sig
# ano_seasonyear <- anosim(PS2_otu_log_dist, PS2_data$SeasonYear)
# ano_seasonyear #sig
# ano_month <- anosim(PS2_otu_log_dist, PS2_data$Month)
# ano_month #sig
# ano_pH <- anosim(PS2_otu_log_dist, PS2_data$pH)
# ano_pH #slightly sig
# ano_watertemp <- anosim(PS2_otu_log_dist, PS2_data$Watertemp_C)
# ano_watertemp #not sig

#####on bray distance original data! - used this, because most significant diff
### PERMANOVA - hypothesis testing
PS2_otu <- as.matrix(otu_table(PS2_taxtree))
PS2_otu_dist <- phyloseq::distance(PS2_otu, method = "bray")
PS2_data <- data.frame(sample_data(PS2_taxtree))

adonis_season2 <- adonis(PS2_otu_dist ~ Season + SeasonYear + Month + pH + pH_range + Watertemp_C, data = PS2_data)
adonis_season2 #season and seasonyear are very significant, month & pH range sig

## anosim - analysis of similarities - test if env factors have influence on community composition
ano_season2 <- anosim(PS2_otu_dist, PS2_data$Season)
ano_season2 #sig
ano_seasonyear2 <- anosim(PS2_otu_dist, PS2_data$SeasonYear)
ano_seasonyear2 #sig
ano_month2 <- anosim(PS2_otu_dist, PS2_data$Month)
ano_month2 #sig
ano_pH2 <- anosim(PS2_otu_dist, PS2_data$pH)
ano_pH2 #slightly sig
ano_pHrange2 <- anosim(PS2_otu_dist, PS2_data$pH_range)
ano_pHrange2
ano_watertemp2 <- anosim(PS2_otu_dist, PS2_data$Watertemp_C)
ano_watertemp2 #not sig

## homogeneity of dispersion test (nonmetric based on permutations)
dis_season <- betadisper(PS2_otu_dist, PS2_data$Season)
permutest(dis_season) #slightly sig
dis_seasonyear <- betadisper(PS2_otu_dist, PS2_data$SeasonYear)
permutest(dis_seasonyear)
dis_pH <- betadisper(PS2_otu_dist, PS2_data$pH)
permutest(dis_pH) # sig

### Tukey's Honest Significant Differences
HSD_season <- TukeyHSD(dis_season)
HSD_season

par(mfrow=c(1,2))
plot(HSD_season, las=1)
plot(dis_season, las=1)
HSD_seasonyear <- TukeyHSD(dis_seasonyear)
HSD_seasonyear
par(mfrow=c(1,2))
plot(HSD_seasonyear, las=1)
plot(dis_seasonyear, las=1)
HSD_pH <- TukeyHSD(dis_pH)
HSD_pH
par(mfrow=c(1,2))
plot(HSD_pH, las=1)
plot(dis_pH, las=1)

#####on wunifrac distance original data!
### PERMANOVA - hypothesis testing

# PS2_otu_wunifrac <- phyloseq::distance(PS2, method = "unifrac", weighted=T)
# PS2_data <- data.frame(sample_data(PS2))
# 
# adonis_season3 <- adonis(PS2_otu_wunifrac ~ Season + SeasonYear + Month + pH + Watertemp_C, data = PS2_data)
# adonis_season3 #season very significant, seasonyear and month slightly
# 
# ## anosim - analysis of similarities - test if env factors have influence on community composition
# ano_season3 <- anosim(PS2_otu_wunifrac, PS2_data$Season)
# ano_season3 #sig
# ano_seasonyear3 <- anosim(PS2_otu_wunifrac, PS2_data$SeasonYear)
# ano_seasonyear3 #sig
# ano_month3 <- anosim(PS2_otu_wunifrac, PS2_data$Month)
# ano_month3 #sig
# ano_pH3 <- anosim(PS2_otu_wunifrac, PS2_data$pH)
# ano_pH3 #slightly sig
# ano_watertemp3 <- anosim(PS2_otu_wunifrac, PS2_data$Watertemp_C)
# ano_watertemp3 #not sig

# generate a biplot with environmental data
# use only OTUs with mean RA above 0.01%
PS2_abundant <- filter_taxa(PS2_taxtree, function(x) mean(x) > 0.01, prune = TRUE)
otu_table_abundant <- otu_table(PS2_abundant)
sample_data_abundant <- data.frame(sample_data(PS2_abundant))
#evaluate whether to use CCA or RDA with Detrended correspondence analysis (DCA)
dca <- decorana(otu_table_abundant)
#check axis lengths of DCA1 
# if <3 use RDA, >4 use CCA, 3-4 use either CCA or RDA
dca
# CCA
otu_table_abundant <- as.data.frame(otu_table_abundant)
# change rows and columns of df
otu_table_trans <- data.frame(t(otu_table_abundant))
#remove columns with unnecessary information from df
#sample_data_abundant <- data.frame(sample_data_abundant[,-c(1,2,7,11,13,14)])
# ordinate
#ord_all <- cca(otu_table_trans ~ Season + SeasonYear + pH + Watertemp_C, sample_data_abundant)
#test without watertemp
ord_all <- cca(otu_table_trans ~ Season + SeasonYear + pH, sample_data_abundant)
plot(ord_all)
plot(ord_all, type="n", scaling="sites")
text(ord_all, scaling="sites")
points(ord_all, display = "sites", scaling="sites", col="blue", bg="red")

#define colors for SeasonYear
colvec <- c("darkseagreen3","deeppink3","goldenrod1","deepskyblue3","darkseagreen4","deeppink4","goldenrod3","deepskyblue4")
#sample_data_abundant$SeasonYear <- as.factor(sample_data_abundant$SeasonYear)
#levels(sample_data_abundant$SeasonYear)

fit <- envfit(ord_all ~ Season + SeasonYear + pH + Watertemp_C, sample_data_abundant, perm=999,display="lc")
#test without watertemp
#fit <- envfit(ord_all ~ Season + SeasonYear + pH, sample_data_abundant, perm=999,display="lc")
plot(ord_all, type="n")
plot(fit)
#points(ord_all)
with(sample_data_abundant, points(ord_all, col =colvec[SeasonYear],pch=21, scaling = "sites", bg = colvec[SeasonYear]))
with(sample_data_abundant,legend("topright", legend=levels(SeasonYear), col=colvec, pch=21, pt.bg=colvec))

#check variance inflation factors VIF - how much is variance inflated by presence of other covariates
#VIF >= 20 strong colinearity, >=10 of concern --remove all above 10
temp <- vif.cca(ord_all)
temp
#should remove watertemp, but haven't!
fit$vectors # pH and temp are very significant

spp.scrs <- as.data.frame(scores(fit, display = "vectors"))
pval <- fit$vectors$pvals
fdat <- cbind(spp.scrs, Vector = rownames(spp.scrs), pval)
#select best env variables based on p-value
bestEnvVar <- rownames(fdat)[fdat$pval<=0.05]
#rem na
#bestEnvVar<-bestEnvVar[!is.na(bestEnvVar)]

# redo CCA using the best env variables
eval(parse(text=paste("ord1 <- cca(otu_table_trans ~ ",do.call(paste,c(as.list(bestEnvVar),sep=" + ")),",data=sample_data_abundant)",sep="")))
summary(ord1)
fit1 <- envfit(ord1,sample_data_abundant[,bestEnvVar], perm = 999, display = "lc", scaling = "sites")

pdf("CCAplot_chl.pdf")
plot(ord1, type="n", xlim = c(-3, 2), ylim = c(-2.5, 1.5)) 
with(sample_data_abundant, points(ord_all, col =colvec[SeasonYear],pch=21, cex=1.5, scaling = "sites", bg = colvec[SeasonYear]))
plot(fit1, col = "#7e7e7e", cex=1, axis=TRUE, p.max = 0.05)
with(sample_data_abundant,legend(x = "bottomleft", legend=levels(SeasonYear), col=colvec, pch=21, pt.bg=colvec)) 

dev.off()

# plot "fit" - without waterT
scl <- 3 #scaling = 3
plot(ord_all, type="n", scaling = scl)
with(sample_data_abundant, points(ord_all, col =colvec[SeasonYear],pch=21, scaling = "sites", bg = colvec[SeasonYear]))
plot(fit, col = "grey", cex=1.2, axis=TRUE, p.max = 0.05)
with(sample_data_abundant,legend("topright", inset=c(-0.3,0),legend=levels(SeasonYear), col=colvec, pch=21, pt.bg=colvec))


#library(dendextend)

#Check for differences in families in seasons
# Agglomerate to family and rename
PS2_fam_taxtree <- phyloseq::tax_glom(PS2_taxtree, "Family")
#replace taxa names by family names
phyloseq::taxa_names(PS2_fam_taxtree) <- phyloseq::tax_table(PS2_fam_taxtree)[, "Family"]

#Melt and plot
phyloseq::psmelt(PS2_fam_taxtree) %>%
ggplot(data = ., aes(x = Season, y = Abundance, fill = Family)) +
  geom_boxplot(outlier.shape  = NA) +
  scale_fill_manual(colour = my_color2) +
  #geom_jitter(aes(color = Family), height = 0, width = .2) +
  geom_jitter() +
  labs(x = "", y = "Abundance\n") +
  facet_wrap(~ Family, scales = "free")

# other way
 p = plot_bar(PS2_fam_taxtree, "Family", fill="Family", facet_grid=Season~pH)
 p + geom_bar(aes(color=Family, fill=Family), stat="identity", position="stack")

# NMDS plot
#dist = phyloseq::distance(PS2, method="jaccard", binary = T)
dist = phyloseq::distance(PS2_taxtree, method="wunifrac")
ordination = ordinate(PS2_taxtree, method="NMDS", distance=dist)
plot_ordination(PS2_taxtree, ordination, color="Season") + 
  theme_classic() +
  stat_ellipse(aes(group = Season)) +
  theme(strip.background = element_blank())
# bray distance same result as jaccard, wunifrac more clear?

# ANOSIM to test
anosim(dist, sample_data(PS2_taxtree)$Season)
#do pairwise ANOSIM?

# double principal coordinates analysis - phylogenetic ordination providing biplot of samples and taxonomic categories
out.dpcoa.log <- ordinate(pslog, method = "DPCoA")
evals <- out.dpcoa.log$eig
# incorporates phylogenetic information, but dominated by first axis
p_dpcoa_log1 <- plot_ordination(pslog, out.dpcoa.log, color = "Season") +
  labs(col = "Season")+
  coord_fixed(sqrt(evals[2] / evals[1]))
# showing taxa that are responsible for axis 1 and 2
p_dpcoa_log2 <- plot_ordination(pslog, out.dpcoa.log, type = "species", color = "Family") +
  coord_fixed(sqrt(evals[2] / evals[1]))
#tried also with clr, but negative values in psclr don't allow this method!
grid.arrange(p_dpcoa_log1, p_dpcoa_log2, ncol=2)

# do a canonical correspondence analysis - including also other environmental factors - pH and water temp
ps_ccpna <- ordinate(pslog, "CCA", formula = pslog ~ Season + SeasonYear + pH + Watertemp_C)
library(ggrepel)
ps_scores <- vegan::scores(ps_ccpna)
sites <- data.frame(ps_scores$sites)
sites <- merge(sites, sample_data(pslog), by = 0)

species <- data.frame(ps_scores$species)
species <- merge(species, tax_table(PS2), by = 0)
evals_prop <- 100 * ps_ccpna$CCA$eig[1:2] / sum(ps_ccpna$CA$eig)
# ggplot() +
#    geom_point(data = sites, aes(x = CCA1, y = CCA2), shape = 2, alpha = 0.5) +
#    geom_point(data = species, aes(x = CCA1, y = CCA2, col = Family), size = 0.5) +
#   geom_text_repel(data = species,
#                      aes(x = CCA1, y = CCA2 %>% filter(CCA2 < "-2"), label = Family),
#             size = 1.5, segment.size = 0.1) + #%>% filter(CCA2 < -2)
# facet_grid(. ~ Season) +
# guides(col = guide_legend(override.aes = list(size = 3))) +
#    labs(x = sprintf("Axis1 [%s%% variance]", round(evals_prop[1], 2)),
#          y = sprintf("Axis2 [%s%% variance]", round(evals_prop[2], 2))) +
#    #scale_color_brewer(palette = "Set2") +
#    scale_color_manual(values= colorRampPalette(brewer.pal(9,"Set1"))(14)) +
#    coord_fixed(sqrt(ps_ccpna$CCA$eig[2] / ps_ccpna$CCA$eig[1])*0.45   ) +
#    theme(panel.border = element_rect(color = "#787878", fill = alpha("white", 0)))


```
```{r, alluvial plot, include = TRUE}

library(ggalluvial)

#use PS2_taxtree_ab dataframe as input
# abs ab
ggplot(PS2_taxtree_ab, aes(x = Timepoint, stratum = Family, alluvium = OTU, y = absnr, fill = Family, label = Family)) +
  geom_flow() +
  geom_stratum(aes(alpha = .5)) +
  geom_alluvium(curve_type = "cubic") +
  #scale_y_log10() +
  scale_fill_manual(values = my_color2) +
  scale_x_discrete(limits = rev) +
  theme_bw()
ggsave("alluvial_chl_abs.png", width = 15, height = 10)
  
# ra
ggplot(PS2_taxtree_ab, aes(x = Timepoint, stratum = Family, alluvium = OTU, y = Abundance, fill = Family, label = Family)) +
  geom_flow() +
  geom_alluvium(curve_type = "cubic") +
  scale_x_discrete(limits = rev) +
  geom_stratum(alpha = .5) +
  scale_fill_manual(values = my_color2) +
  theme_bw()

ggsave("alluvial_chl_RA.png", width = 15, height = 10)

######
PS2_taxtree_ab$SeasonYear <- as.factor(PS2_taxtree_ab$SeasonYear)
# summarize absolute numbers by SeasonYear
PS2_season_ra <- PS2_taxtree_ab %>% group_by(SeasonYear, Family) %>% dplyr::summarize(absnr = mean(absnr)) %>% as.data.frame()
#check for alluvial format
is_alluvia_form(PS2_season_ra, x = "SeasonYear", stratum = "Family", alluvium = "Family", y = "absnr")

ggplot(PS2_season_ra, aes(x = SeasonYear, stratum = Family, alluvium = Family, y= absnr, fill = Family, label = Family)) + 
  geom_flow() +
  #scale_x_discrete(limits = rev) +
  #scale_y_log10() +
  geom_stratum(alpha = 0.5) +
  geom_alluvium(alpha = 0.5) +
  scale_fill_manual(values = my_color2) +
  theme_bw() +
  xlab("") +
  ylab("mean 16S rRNA gene copies/g sediment") +
  theme(axis.text = element_text(size = 16)) +
  theme(axis.title = element_text(size =20))

ggsave("alluvial_chl_seasonYear_mean.png", width = 15, height = 10)

#### alluvial plot for summer TP for chl families & protist classes
#subset chlamydiae for summer
PS2_taxtree_summer <- subset.data.frame(PS2_taxtree_ab, Season == "summer")
Chl <- PS2_taxtree_summer %>% group_by(Timepoint, Family) %>% dplyr::summarize(RA = sum(Abundance)) %>% as.data.frame()

# import table for unicellular eukaryotes
PS_U_ra <- read.table("PS_U_ra.tsv", sep="\t", header = TRUE)

# subset summer 
PS_U_summer <- subset.data.frame(PS_U_ra, Season == "summer")
PS_U_summer$Abundance <- as.numeric(PS_U_summer$Abundance)
is.numeric(PS_U_summer$Abundance)
#replace NA by 0
PS_U_summer[is.na(PS_U_summer)] <- 0

Prot <- PS_U_summer %>% group_by(Timepoint, Division) %>% dplyr::summarize(RA_prot = sum(Abundance)) %>% as.data.frame()

# merge dataframes by timepoint
Chl_Prot <- merge(Chl, Prot, by = "Timepoint", all = TRUE)

ggplot(Chl_Prot, aes(y = freq, axis1 = Family, axis2 = Timepoint, axis3 = Division)) +
  geom_alluvium(aes(fill = Abundance)) +
  geom_stratum +
  theme_bw()


```


